### ISSUE TITLE
Checkline exploit

### ISSUE THREAD

**Issue by Alayan-stk-2** [supertuxkart/stk-code#3161]
See this video as an example of possible abuse : https://www.youtube.com/watch?v=5Ckbd1VL0-o

It can be technically solved by adding a lot of checklines to all tracks, but that is not an elegant solution and can't be expected of addons track anyway. This would also be forgotten from time to time, leading to a maintenance burden.

So another solution has to be put in place.

Would there be bad side effects if the lap validation was blocked unless the kart drove a minimum of X% of the track's distance ? With a number like 75%, I think it should always cover players taking the shortest legal path while killing the abuse, as it requires driving in reverse.

(Note : I've not checked exactly how the distance the kart did is calculated. The implementation of a fix depends heavily on how this number is calculated, so the above suggested solution may not work)

**Comment by auriamg** [comment]
Maybe, but really we usually just want tracks to have at least 2 checklines in general, since that fixes many issues (prevents shortcuts too). So not sure a code solution is really needed, probably we only have a couple tracks that need one more driveline and that would be easily fixed that way

**Comment by Alayan-stk-2** [comment]
Tracks having multiple checklines do not always fix some kind of abuses.

See XR591 : https://www.youtube.com/watch?v=hUu7NqYUqFk

You can see the original title mentioned a "rescue abuse". That is because driving backward, it's possible to get rescued and sent back to the start.

See : https://www.youtube.com/watch?v=C681Cs-u7fQ

I removed it from the title and post because I figured out that a fix to this exploit shouldn't involve modifying the rescue behavior, but rather tackle the root cause (driving backwards and still getting the lap counted).

Finally, my personal observation is that relying on tracks being full-proof will leave holes open. I could talk about addon tracks, but the fact is that numerous tracks of the main game are still affected after years and years. I'm not confident they will all be patched and that this will never be forgotten for any future track.

**Comment by auriamg** [comment]
Somewhat agreed, but then the XR591 exploit you've linked to would not be fixed by your proposed solution, so you've kind of demonstrated how limited your proposed solution is :)

We usually have this stance on shortcuts : we fix the worst ones, and for the others, if players are clever and skilled enough to find other shortcuts, let them have it they deserve it, if that's how they enjoy playing the game

**Comment by Alayan-stk-2** [comment]
The XR591 exploit may not be fixed - it would probably better be solved at the track level ; but the proposed solution would at least fix all "drive in the wrong direction" exploits, all "go out of bounds and be rescued" exploits (like in Fort Magma) and at least some of the most blatant other exploits (like the one in Scotland over the river).

**Comment by auriamg** [comment]
Green valley and Nessie's Pond were fixed by checklines.

Look, your idea of a percentage of track length is not necessarily bad, but with multiple paths, cannons, possibly eventually scripting, I expect lots of weird edge cases. We used to have shortcut detection in previous versions of STK that constantly had false positives and false negatives no matter how hard we tweaked it; that's when we decided to stop the trouble, remove automated shortcut detection and just use checkpoints, which require like 10 minutes to place and are much more simple to get working correctly. So I expect we'll just keep using those good old reliable and simple checkpoints.

FInally, for some like XR591, I think it's okay if created/skilled players can do little shortcuts to shave off a few seconds, so long as they don't skip too big parts of the track.
